File: 2023/07.17/송기훈/ABCDE.java
package baekjoon.gold.five;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class ABCDE {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int N, M;
    static ArrayList<ArrayList<Integer>> nodeList;
    static boolean flag;
    static boolean[] visited;


    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        nodeList = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            nodeList.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            nodeList.get(a).add(b);
            nodeList.get(b).add(a);
        }

        for (int i = 0; i < N; i++) {
            visited = new boolean[N];
            visited[i] = true;
            dfs(i, 0);
            visited[i] = false;
            if (flag) {
                break;
            }
        }

        if (flag) {
            System.out.println(1);
        } else {
            System.out.println(0);
        }
    }

    static void dfs(int index, int depth) {
        if (depth == 4) {
            flag = true;
            return;
        }

        ArrayList<Integer> tempList = nodeList.get(index);
        for (int tempIndex : tempList) {
            if (!visited[tempIndex]) {
                visited[tempIndex] = true;
                dfs(tempIndex, depth + 1);
                visited[tempIndex] = false;
            }
        }
    }

}

File: 2023/07.17/조수연/ABCDE.kt
import java.io.*
import java.util.*
import kotlin.system.exitProcess

lateinit var graph: ArrayList<ArrayList<Int>>
lateinit var visited: BooleanArray

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val (n, m) = readLine().trim().split(" ").map { it.toInt() }

    graph = ArrayList()

    for (i in 0 until n) {
        graph.add(ArrayList())
    }

    visited = BooleanArray(n)

    repeat(m) {
        val token = StringTokenizer(readLine())
        val s = token.nextToken().toInt()
        val e = token.nextToken().toInt()

        graph[s].add(e)
        graph[e].add(s)
    }

    for (i in 0 until n) {
        visited = BooleanArray(n)
        (dfs(i, 0))
    }

    bw.write("0")


    bw.flush()
    bw.close()
}

fun dfs(start: Int, depth: Int): Boolean {
    if (depth == 4) {
        print(1)
        exitProcess(0)
    }

    visited[start] = true

    for (next in graph[start]) {
        if (!visited[next]) {
            dfs(next, depth + 1)
            visited[next] = false
        }
    }

    return false
}